[Aakansha Narain] iP

aakanshanarain opened the pull request:
### Duke: Your personal task manager

Welcome to Duke! Duke provides an interface to keep track of your tasks to make sure you never fall back on work! Here's some stuff you can do with Duke:

- Store _todo_, _event_ and _deadline_ tasks
- Add date and time to your event and deadline tasks
- Tag your tasks to group tasks together

To use Duke, all you have to do is:

1. Download the program from [here](https://github.com/aakanshanarain/ip) 
2. Double click it
3. Add your tasks
4. Let it do the rest for you!

An example from our previous users:

- [ ] Finish CSXXXX Assignment (by: Sep 20 2021, 11:59pm) 
- [ ] Meet friends (at: Sep 21 2021, 3:00pm)
- [ ] Buy groceries
- [ ] Attend book club meeting (at: Sep 25, 5:00pm) #bookClub
- [ ] Write book report #bookClub

Take a look at some reviews from current users:

> Duke has helped me manage my tasks so much more efficiently! It really is a second brain!

> Ever since I started using Duke, I've never missed a deadline! It really helps me through the stressful university semester!

For more users experienced in programming, play around with the software to make it more customisable! Here's the `main` method to get you started:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```


**Get started with Duke now! 🤩**

Commit:
Add Gradle support

aakanshanarain committed:
Level-1

aakanshanarain committed:
Level-2

aakanshanarain committed:
Level-3

aakanshanarain committed:
Level-4

aakanshanarain committed:
A-TextUiTesting

aakanshanarain committed:
Level-5

aakanshanarain committed:
Level-6

aakanshanarain committed:
Level-6

aakanshanarain committed:
Add more OOP

aakanshanarain committed:
Move classes to duke package

yuifuku1118 commented on src/main/java/duke/EventInput.java:
```diff
@@ -0,0 +1,27 @@
+        if (input.length() == 5) {
+            throw new EmptyDescriptionException("error" );
+        }
+        int charIndex = input.indexOf("/" );
+        int atIndex = charIndex + 4;
```
I like how you named your variables here! (Easy to understand)

yuifuku1118 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,28 @@
+
+    public void add (Task task) {
+        todoList.add(task);
+    }
```
I like how you named your variables with verbs, easy to understand at one glance.

yuifuku1118 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,17 @@
+public class Event extends Task {
+
+    private String at;
+
```
You could remove the extra empty line

yuifuku1118 commented on src/main/java/duke/InputHandler.java:
```diff
@@ -0,0 +1,14 @@
+package duke;
+
+public abstract class InputHandler {
+    protected Ui ui;
+    protected TaskList taskList;
```
You could add a line between? (To keep your coding style consistent, same thing with couple of other files)

yuifuku1118 reviewed (commented):
LGTM!! 
Overall, your code is nice and well organized. 
Just added some comment and possible improvements to make.

vigneshsankariyer1234567890 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+//Solution below slightly adapted from https://github.com/Wincenttjoi/CS2103T-duke-chatbot/blob/master/src/main/java/duke/Duke.java
+
+import duke.*;
```
Should you be including the individual packages here rather than a wildcard?

vigneshsankariyer1234567890 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+
+    private Ui ui;
+    private TaskList taskList;
+    private boolean exit;
```
You could change this to a more boolean sounding name!

vigneshsankariyer1234567890 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+    }
+
+    public enum InputCommands {
+        bye, list, done, delete, todo, deadline, event
```
Since these are going to be unchanged, you could possibly rename to be capitalised? For example, BYE, LIST, DONE etc. However, i like the naming of the enum InputCommands here!

vigneshsankariyer1234567890 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+
+import duke.*;
+
 public class Duke {
```
You could also add some comments here! I think this would allow you to communicate your code effectively!

vigneshsankariyer1234567890 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+        bye, list, done, delete, todo, deadline, event
+    }
+
+    public void start () {
```
I think you do not need to add a space after the method name here! just start() would do

vigneshsankariyer1234567890 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        new Duke().start();
```
i think you could explicitly close Duke after this by adding a System.close() here!

vigneshsankariyer1234567890 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,16 @@
+
+public class Deadline extends Task {
+
+    protected String by;
```
i think this variable should be private even if it is only going to be accessed within the package! you could add a getter to retrieve this variable

vigneshsankariyer1234567890 commented on src/main/java/duke/DeadlineInput.java:
```diff
@@ -0,0 +1,24 @@
+        super(ui, taskList);
+    }
+
+    @Override
```
I think you could refactor your toString method here! or possibly store the instantiated objects as variables instead!

vigneshsankariyer1234567890 commented on src/main/java/duke/DeleteInput.java:
```diff
@@ -0,0 +1,23 @@
+            throw new EmptyDescriptionException("error");
+        }
+
+        char taskIndex = input.charAt(7);
```
i like this separation of lines. allows me to understand the flow of logic slightly better.

vigneshsankariyer1234567890 commented on src/main/java/duke/DeleteInput.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public String handle(String input) throws EmptyDescriptionException {
+        if (input.length() == 6) {
```
I'm not too sure whether i understand the reasoning behind using length() here. Could you comment on this?

vigneshsankariyer1234567890 commented on src/main/java/duke/DoneInput.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public String handle (String input) throws EmptyDescriptionException {
+        if (input.length() == 4) {
+            throw new EmptyDescriptionException("error");
```
I think it's great to have different Exceptions here! really improves the OO-ness of the project.

vigneshsankariyer1234567890 commented on src/main/java/duke/DoneInput.java:
```diff
@@ -0,0 +1,24 @@
+            throw new EmptyDescriptionException("error");
+        }
+
+        char taskIndex = input.charAt(5);
```
I think you could rename the variable here, to something like indexOfTask?

vigneshsankariyer1234567890 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.ArrayList;
+
+public class TaskList {
+    private static ArrayList<Task> todoList;
```
You could rename this to tasks to keep it consistent with the coding standards provided by prof!

vigneshsankariyer1234567890 replied:
this could also be instance level (non-static) rather than class level! I think it would be easier to keep track of the collection this way.

vigneshsankariyer1234567890 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,28 @@
+        return todoList;
+    }
+
+    public void add (Task task) {
```
again, as with other lines of code, you could remove the space between the method name and parameter parentheses

vigneshsankariyer1234567890 reviewed (commented):
Hey Aakansha!

Overall, really liked the organisation of your project. It really adheres to the principles of OO! I'll be leaving some comments on some parts of your project. Once again, well done!

aakanshanarain committed:
Add Storage class

aakanshanarain committed:
Add JUnit Tests

aakanshanarain committed:
Add JAR file

aakanshanarain committed:
Add Javadocs to public methods

aakanshanarain committed:
Tweak to comply with coding standards

aakanshanarain committed:
Add find command functionality

aakanshanarain committed:
Merge branch 'branch-A-CodingStandard'
This merge includes changes to the coding standard

aakanshanarain committed:
Resolve conflicts while merging branch-Level-9

aakanshanarain committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle
This merges the gradle-support to implement Scenario 2 while adding Gradle to the iP

aakanshanarain committed:
Add Gradle support to automate some build tasks

aakanshanarain committed:
Add GUI to Duke

aakanshanarain committed:
Add storage functionality to save tasklist

aakanshanarain committed:
Add date and time conversion for task inputs

aakanshanarain committed:
Add more OOP

aakanshanarain committed:
Add more JavaDocs and tweak coding style

aakanshanarain committed:
Add assertions

aakanshanarain committed:
Revert "Add assertions"

This reverts commit c228d14ec3566a06778019f71afd8b17ce664681.

Reverting commit to be made in another branch

aakanshanarain committed:
Add assertions

aakanshanarain committed:
Improve overall code quality: modify to adhere to SLAP and add better comments

The Storage class had code at varying levels of abstraction. It had code that was responsible for parsing the stored tasks, as well as loading them into the UI at the same time.

Let's move the code responsible for parsing into the Parser class to eliminate that level of abstraction from the Storage class, keeping it solely responsible for storing and loading tasks to and from the storage file.

The Parser class already deals with all the relevant parsing needed to read user inputs and other text formats, so this maintains abstraction well. It also makes the code more readable as there is no convoluted parsing to read while looking at the code to load tasks to the UI.

aakanshanarain committed:
Merge pull request #2 from aakanshanarain/branch-A-Assertions

Add assertions to some fragments of code

aakanshanarain committed:
Merge pull request #3 from aakanshanarain/branch-A-CodeQuality

Modifications to the code quality

aakanshanarain committed:
Add tagging functionality for tasks

Tagging can be useful to keep track of similar tasks in the task list.

Let's add a tag command word and accordingly add methods to the Parser and UI classes, as well as make changes to the toString methods to make the UI messages more readable for the user in the case they add a tag to the tasks.

In keeping with the functions of the classes, the Parser parses the tag command and then the CommandHandler ensures the actual handling of the command. The UI class then returns the output message. This is in accordance to the levels of abstraction and classification established.

aakanshanarain committed:
Rectify problems with the GUI

aakanshanarain committed:
Add Ui.png

aakanshanarain committed:
Add User Guide

aakanshanarain committed:
Change homepage screenshot for user guide

aakanshanarain committed:
Set theme jekyll-theme-minimal

aakanshanarain committed:
Set theme jekyll-theme-minimal

aakanshanarain committed:
Set theme jekyll-theme-cayman

aakanshanarain committed:
Set theme jekyll-theme-minimal

aakanshanarain committed:
Add acknowledgments

aakanshanarain committed:
Merge branch 'master' of github.com:aakanshanarain/ip
Add acknowledgements

aakanshanarain committed:
Tweak README.md

aakanshanarain committed:
Update README.md

aakanshanarain committed:
Set theme jekyll-theme-minimal

aakanshanarain committed:
Update README.md

aakanshanarain committed:
Update README.md

aakanshanarain committed:
Tweak README.md to fix images and replace Ui.png

aakanshanarain committed:
Tweak README.md to fix Homepage image

aakanshanarain committed:
Update to help command for the GUI

aakanshanarain committed:
Tweak Duke.java

aakanshanarain committed:
Clean code by removing redundant methods and add more JavaDocs

aakanshanarain committed:
Delete tasks.txt

aakanshanarain committed:
Fix issues with launching on Windows and Linux

aakanshanarain committed:
Refactor code in CommandHandler, Parser and Ui

aakanshanarain committed:
Update exceptions to be more comprehensive
